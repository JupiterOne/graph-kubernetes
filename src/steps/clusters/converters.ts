import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Cluster } from '@kubernetes/client-node';
import { Entities } from '../constants';

function getClusterKey(id: string) {
  return `kube_cluster:${id}`;
}

export function createClusterEntity(data: Cluster) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CLUSTER._class,
        _type: Entities.CLUSTER._type,
        _key: getClusterKey(data.name),
        name: data.name,
        displayName: data.name,
        server: data.server,
        skipTlsVerify: data.skipTLSVerify,
        caData: data.caData,
      },
    },
  });
}
