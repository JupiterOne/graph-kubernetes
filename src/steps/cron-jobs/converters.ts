import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createCronJobEntity(data: k8s.V1CronJob) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CRONJOB._class,
        _type: Entities.CRONJOB._type,
        _key: data.metadata!.uid!,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        deletedOn: parseTimePropertyValue(data.metadata?.deletionTimestamp),
        // spec fields
        concurrencyPolicy: data.spec?.concurrencyPolicy,
        failedJobsHistoryLimit: data.spec?.failedJobsHistoryLimit,
        schedule: data.spec?.schedule,
        startingDeadlineSeconds: data.spec?.startingDeadlineSeconds,
        successfulJobsHistoryLimit: data.spec?.successfulJobsHistoryLimit,
        suspend: data.spec?.suspend,
        'status.lastScheduleTime': parseTimePropertyValue(
          data.status?.lastScheduleTime,
        ),
      },
    },
  });
}
