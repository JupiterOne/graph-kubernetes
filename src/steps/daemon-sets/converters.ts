import * as k8s from '@kubernetes/client-node';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

export function createDaemonSetEntity(data: k8s.V1DaemonSet) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.DAEMONSET._class,
        _type: Entities.DAEMONSET._type,
        _key: data.metadata!.uid!,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        deletedOn: parseTimePropertyValue(data.metadata?.deletionTimestamp),
        // spec fields
        minReadySeconds: data.spec?.minReadySeconds,
        revisionHistoryLimit: data.spec?.revisionHistoryLimit,
        'strategy.type': data.spec?.updateStrategy?.type,
        'strategy.maxUnavailable':
          data.spec?.updateStrategy?.rollingUpdate?.maxUnavailable,
        'status.collisionCount': data.status?.collisionCount,
        'status.currentNumberScheduled': data.status?.currentNumberScheduled,
        'status.desiredNumberScheduled': data.status?.desiredNumberScheduled,
        'status.numberAvailable': data.status?.numberAvailable,
        'status.numberMisscheduled': data.status?.numberMisscheduled,
        'status.numberReady': data.status?.numberReady,
        'status.numberUnavailable': data.status?.numberUnavailable,
        'status.observedGeneration': data.status?.observedGeneration,
        'status.updatedNumberScheduled': data.status?.updatedNumberScheduled,
      },
    },
  });
}
