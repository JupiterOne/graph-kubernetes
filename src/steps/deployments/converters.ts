import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  generateRelationshipKey,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';
import { RelationshipClass } from '@jupiterone/data-model';
import { V1VolumeMount } from '@kubernetes/client-node';

export function getVolumeKey(
  namespace: string,
  deploymentId: string,
  volumeName: string,
) {
  return `${namespace}/${deploymentId}/${volumeName}`;
}

export function createVolumeEntity(
  namespace: string,
  deploymentUid: string,
  data: k8s.V1Volume,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.VOLUME._class,
        _type: Entities.VOLUME._type,
        _key: getVolumeKey(namespace, deploymentUid, data.name),
        name: data.name,
        displayName: data.name,
        'ephemeral.readOnly': data.ephemeral?.['readOnly'],
        'hostPath.path': data.hostPath?.path,
        'hostPath.type': data.hostPath?.type,
        classification: null,
        encrypted: false,
      },
    },
  });
}

export function getContainerSpecKey(
  namespace: string,
  deploymentId: string,
  specName: string,
) {
  return `${namespace}/${deploymentId}/${specName}`;
}

export function createContainerSpecEntity(
  namespace: string,
  deploymentUid: string,
  data: k8s.V1Container,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CONTAINER_SPEC._class,
        _type: Entities.CONTAINER_SPEC._type,
        _key: getContainerSpecKey(namespace, deploymentUid, data.name),
        name: data.name,
        displayName: data.name,
        image: data.image,
        imagePullPolicy: data.imagePullPolicy,
        'securityContext.allowPrivilegeEscalation':
          data.securityContext?.allowPrivilegeEscalation,
        'securityContext.privileged': data.securityContext?.privileged,
        'securityContext.procMount': data.securityContext?.procMount,
        'securityContext.readOnlyRootFilesystem':
          data.securityContext?.readOnlyRootFilesystem,
        'securityContext.runAsGroup': data.securityContext?.runAsGroup,
        'securityContext.runAsNonRoot': data.securityContext?.runAsNonRoot,
        'securityContext.runAsUser': data.securityContext?.runAsUser,
      },
    },
  });
}

export function createDeploymentEntity(data: k8s.V1Deployment) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.DEPLOYMENT._class,
        _type: Entities.DEPLOYMENT._type,
        // metadata properties
        _key: data.metadata!.uid!,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec properties
        minReadySeconds: data.spec?.minReadySeconds,
        paused: data.spec?.paused,
        progressDeadlineSeconds: data.spec?.progressDeadlineSeconds,
        replicas: data.spec?.replicas,
        revisionHistoryLimit: data.spec?.revisionHistoryLimit,
        'strategy.type': data.spec?.strategy?.type,
        'strategy.maxSurge': data.spec?.strategy?.rollingUpdate?.maxSurge,
        'strategy.maxUnavailable':
          data.spec?.strategy?.rollingUpdate?.maxUnavailable,
        // TODO: something about spec.template sections maybe?
        'status.availableReplicas': data.status?.availableReplicas,
        'status.collisionCount': data.status?.collisionCount,
        'status.observedGeneration': data.status?.observedGeneration,
        'status.readyReplicas': data.status?.readyReplicas,
        'status.replicas': data.status?.replicas,
        'status.unavailableReplicas': data.status?.unavailableReplicas,
        'status.updatedReplicas': data.status?.updatedReplicas,
      },
    },
  });
}

function getUniqueContainerSpecVolumeMountKey(volumeMount: V1VolumeMount) {
  let key = volumeMount.name;

  if (volumeMount.mountPath) {
    key += `/mount_path_${volumeMount.mountPath}`;
  }

  if (volumeMount.subPath) {
    key += `/mount_sub_path_${volumeMount.subPath}`;
  }

  return key;
}

function generateContainerSpecToVolumeRelationshipKey({
  _class,
  containerSpecEntity,
  volumeEntity,
  volumeMount,
}: {
  _class: RelationshipClass;
  containerSpecEntity: Entity;
  volumeEntity: Entity;
  volumeMount: V1VolumeMount;
}) {
  const baseRelationshipKey = generateRelationshipKey(
    _class,
    containerSpecEntity,
    volumeEntity,
  );

  const volumeMountKey = getUniqueContainerSpecVolumeMountKey(volumeMount);
  return `${baseRelationshipKey}/mount_${volumeMountKey}`;
}

export function createContainerSpecToVolumeRelationship({
  containerSpecEntity,
  volumeEntity,
  volumeMount,
}: {
  containerSpecEntity: Entity;
  volumeEntity: Entity;
  volumeMount: V1VolumeMount;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.USES,
    from: containerSpecEntity,
    to: volumeEntity,
    properties: {
      _key: generateContainerSpecToVolumeRelationshipKey({
        _class: RelationshipClass.USES,
        containerSpecEntity,
        volumeEntity,
        volumeMount,
      }),
      readOnly: volumeMount.readOnly,
      mountPath: volumeMount.mountPath,
      mountPropagation: volumeMount.mountPropagation,
      subPath: volumeMount.subPath,
    },
  });
}
