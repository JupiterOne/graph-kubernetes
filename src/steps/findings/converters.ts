import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { Finding } from '.';
import { Entities } from '../constants';

function getFindingKey(version: string, testNumber: string) {
  return `kube_finding:${version}:${testNumber}`;
}

function getSeverity(finding: Finding) {
  // TODO, somehow..
  return 'informational';
}

function getNumericSeverity(finding: Finding) {
  // TODO, somehow..
  return 1;
}

export function createFindingEntity(finding: Finding) {
  return createIntegrationEntity({
    entityData: {
      source: finding,
      assign: {
        _class: Entities.FINDINGS._class,
        _type: Entities.FINDINGS._type,
        _key: getFindingKey(finding.version, finding.testNumber),
        name: `${finding.version}/${finding.testNumber}`,
        displayName: `${finding.version}/${finding.testNumber}`,
        complianceStandardName: 'CIS Kubernetes',
        complianceStandardVersion: finding.version.split('-')[1],
        complianceStandardRequirement: finding.testNumber,
        status: finding.status,
        category: 'benchmark',
        severity: getSeverity(finding),
        numericSeverity: getNumericSeverity(finding),
        open: false,
      },
    },
  });
}
