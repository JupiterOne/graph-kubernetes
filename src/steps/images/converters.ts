import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';
export function createContainerImageEntity(data: k8s.V1ContainerImage) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.IMAGE._class,
        _type: Entities.IMAGE._type,
        _key: data.names![0],
        names: data.names,
        name: data.names![0],
        sizeInBytes: data.sizeBytes,
      },
    },
  });
}
