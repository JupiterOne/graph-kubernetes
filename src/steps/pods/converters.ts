import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';

export function createPodEntity(data: k8s.V1Pod) {
  // TODO: We could remove data.spec.containers array from this (from rawData)
  // TODO: We may want to take the spec.volumes, delete it from here and create new entities/relationships
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.POD._class,
        _type: Entities.POD._type,
        // metadata properties
        _key: data.metadata!.uid!,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec properties
        activeDeadlineSeconds: data.spec?.activeDeadlineSeconds,
        automountServiceAccountToken: data.spec?.automountServiceAccountToken,
        dnsPolicy: data.spec?.dnsPolicy,
        enableServiceLinks: data.spec?.enableServiceLinks,
        hostIPC: data.spec?.hostIPC,
        hostNetwork: data.spec?.hostNetwork,
        hostPID: data.spec?.hostPID,
        hostname: data.spec?.hostname,
        nodeName: data.spec?.nodeName,
        preemptionPolicy: data.spec?.preemptionPolicy,
        priority: data.spec?.priority,
        priorityClassName: data.spec?.priorityClassName,
        restartPolicy: data.spec?.restartPolicy,
        runtimeClassName: data.spec?.runtimeClassName,
        schedulerName: data.spec?.schedulerName,
        serviceAccount: data.spec?.serviceAccount,
        serviceAccountName: data.spec?.serviceAccountName,
        setHostnameAsFQDN: data.spec?.setHostnameAsFQDN,
        shareProcessNamespace: data.spec?.shareProcessNamespace,
        subdomain: data.spec?.subdomain,
        terminationGracePeriodSeconds: data.spec?.terminationGracePeriodSeconds,
        'status.hostIP': data.status?.hostIP,
        'status.message': data.status?.message,
        'status.nominatedNodeName': data.status?.nominatedNodeName,
        'status.phase': data.status?.phase,
        'status.podIP': data.status?.podIP,
        'status.qosClass': data.status?.qosClass,
        'status.reason': data.status?.reason,
        'status.startTime': parseTimePropertyValue(data.status?.startTime),
      },
    },
  });
}

// TODO: better container key?
// 'Container' is not a Kubernetes resource (the smallest resource is a pod),
// so we don't have the uid for a container. We have uid for pod which is container's
// parent so we could craft a key using it somehow
export function getContainerKey(name: string) {
  return `container:${name}`;
}

// TODO: this isn't pod, but if we're going to break out container data
// off of the pod resource, we may not want to create a different folder just
// to house this converter, or do we?
export function createContainerEntity(data: k8s.V1Container) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.CONTAINER._class,
        _type: Entities.CONTAINER._type,
        // minLength issue
        _key: getContainerKey(data.name),
        name: data.name,
        displayName: data.name,
        image: data.image,
        imagePullPolicy: data.imagePullPolicy,
        terminationMessagePath: data.terminationMessagePath,
        terminationMessagePolicy: data.terminationMessagePolicy,
      },
    },
  });
}
