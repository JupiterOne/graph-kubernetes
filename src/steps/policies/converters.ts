import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import * as k8s from '@kubernetes/client-node';
import { Entities } from '../constants';

export function createNetworkPolicyEntity(data: k8s.V1NetworkPolicy) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.NETWORK_POLICY._class,
        _type: Entities.NETWORK_POLICY._type,
        // metadata properties
        _key: data.metadata!.uid!,
        name: data.metadata?.name,
        displayName: data.metadata?.name,
        generation: data.metadata?.generation,
        deletionGracePeriodSeconds: data.metadata?.deletionGracePeriodSeconds,
        resourceVersion: data.metadata?.resourceVersion,
        createdOn: parseTimePropertyValue(data.metadata?.creationTimestamp),
        // spec
        // TODO: get egress data (mapping? separate entities? combination?)
        // TODO: get ingress data (mapping? separate entities? combination?)
        policyTypes: data.spec?.policyTypes,
      },
    },
  });
}
