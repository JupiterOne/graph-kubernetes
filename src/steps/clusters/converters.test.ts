import { Cluster } from '@kubernetes/client-node';
import { createClusterEntity } from './converters';

describe('#createClusterEntity', () => {
  test('should convert data', () => {
    const cluster: Cluster = {
      name: 'dummy-data',
      server: 'dummy-data',
      skipTLSVerify: false,
    };
    expect(createClusterEntity(cluster)).toMatchSnapshot();
  });
});
