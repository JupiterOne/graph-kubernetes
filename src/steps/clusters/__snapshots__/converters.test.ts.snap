// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createClusterEntity should convert data 1`] = `
{
  "_class": [
    "Cluster",
  ],
  "_key": "kube_cluster:dummy-data",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "name": "dummy-data",
        "server": "dummy-data",
        "skipTLSVerify": false,
      },
    },
  ],
  "_type": "kube_cluster",
  "createdOn": undefined,
  "displayName": "dummy-data",
  "name": "dummy-data",
  "server": "dummy-data",
  "skipTlsVerify": false,
}
`;
